### PR TITLE
Disallow inferring array sizes from the contexts that are not permitted

### LINKED ISSUE #34017: Infer array sizes from the context
**Description:**
According to the spec discussion (https://github.com/ballerina-platform/ballerina-spec/issues/993), array size can be inferred only from the array constructors. But the current implementation supports beyond this requirement and we have to disallow them.

### PR DESCRIPTION
**Pull request by SandaruJayawardana**
Fixes #34017

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java [hunk 1]
```diff
@@ -3899,7 +3899,12 @@ public void visit(BLangWorkerSend workerSendNode) {
 
     @Override
     public void visit(BLangReturn returnNode) {
-        this.typeChecker.checkExpr(returnNode.expr, this.env, this.env.enclInvokable.returnTypeNode.getBType());
+        BType returnType = this.env.enclInvokable.returnTypeNode.getBType();
+        Location pos = this.typeChecker.getLocationOfInferredArray(returnType);
+        if (pos != null) {
+            dlog.error(pos, DiagnosticErrorCode.CLOSED_ARRAY_TYPE_CAN_NOT_INFER_SIZE);
+        }
```
[the remark this answers is not in the document]
Comment (chiranSachintha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:3906]: Can't we refactor this logic in the following way?

1. When resolving the symbol of the array type node(In the visitor of `BLangArrayType` node).

- We can check whether the array type has an open sealed dimension or not
- then check the node in which has array type.
- according to that we can give the error.

IMO by doing this we didn't need to fix each specific case

**Comment by chiranSachintha** [comment]
According to the spec change we need to disallow both these cases(still allow with this change)
```ballerina
// Example 1
public function main() {
    map<int[1][*]> a = {};
}

// Example 2
function value(int[*] x = []) {

}
```
@hasithaa is there any special reason to disallow second example?

**Comment by SandaruJayawardana** [comment]
> is there any special reason to disallow second example?

IMO, there is no effective usage since this `x` is need to be a fixed size. Further, it can causes some conflicts, because size of the `x` is dependent on the argument of first function call.

**Comment by gimantha** [comment]
Shall we change the base to 2.0.x?

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/arrays/sealed_array_semantics_negative.bal [hunk 2]
```diff
@@ -113,3 +113,24 @@ function openToSealArrrayAssignment() {
     int[] intArr1 = [];
     int[*] intArr2 = intArr1.clone();
 }
+
```
Comment (KRVPerera) [inline tests/jballerina-unit-test/src/test/resources/test-src/statements/arrays/sealed_array_semantics_negative.bal:116]: I believe we need to consider 
```Ballerina
type ARR int[*];
```

Although I believe example 2 is okay
```Ballerina
// Example 2
function value(int[*] x = []) {}
```
@chiranSachintha Cant we think of the example 2 as having zero members

**Comment by SandaruJayawardana** [comment]
Open [new PR](https://github.com/ballerina-platform/ballerina-lang/pull/34612) with the review suggestions and ballerina-2.0.x as the base.

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/statements/arrays/SealedArrayTest.java [hunk 3]
```diff
@@ -291,8 +291,26 @@ public void testSemanticsNegativeSealedArrays() {
                         "'FiniteFour'", 108, 20);
         BAssertUtil.validateError(semanticsNegative, i++, "invalid list member access expression: value space " +
                 "'FiniteFive' out of range", 109, 23);
-        BAssertUtil.validateError(semanticsNegative, i, "incompatible types: expected 'int[*]', found " +
-                        "'int[]'", 114, 22);
+        BAssertUtil.validateError(semanticsNegative, i++, "invalid usage of closed type: can not infer array size",
```
[the remark this answers is not in the document]
Comment (SandaruJayawardana) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/statements/arrays/SealedArrayTest.java:294]: +1

[the remark this answers is not in the document]
Comment (SandaruJayawardana) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/statements/arrays/SealedArrayTest.java:294]: `invalid usage of inferred-array-length: can not infer array size` is this okay?
or is it okay to say just `can not infer array size`?
@KRVPerera

[the remark this answers is not in the document]
Comment (SandaruJayawardana) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/statements/arrays/SealedArrayTest.java:294]: `invalid usage of inferred-array-length: can not infer array size` is this okay?
or is it okay to say just `can not infer array size`?
@KRVPerera

**Pull request closed without merging**